core/src/main/scala/kafka/server/MetadataCache.scala

```diff
@@ -302,22 +302,32 @@ class MetadataCache(brokerId: Int) extends Logging {
           copy ++= oldPartitionStates
           partitionStates += (topic -> copy)
         }
-        updateMetadataRequest.partitionStates.asScala.foreach { info =>
-          val controllerId = updateMetadataRequest.controllerId
-          val controllerEpoch = updateMetadataRequest.controllerEpoch
-          val tp = new TopicPartition(info.topicName, info.partitionIndex)
-          if (info.leader == LeaderAndIsr.LeaderDuringDelete) {
+
+        val traceEnabled = stateChangeLogger.isTraceEnabled
+        val controllerId = updateMetadataRequest.controllerId
+        val controllerEpoch = updateMetadataRequest.controllerEpoch
+        val newStates = updateMetadataRequest.partitionStates.asScala
+        newStates.foreach { state =>
+          val tp = new TopicPartition(state.topicName, state.partitionIndex)
+          if (state.leader == LeaderAndIsr.LeaderDuringDelete) {
             removePartitionInfo(partitionStates, tp.topic, tp.partition)
-            stateChangeLogger.trace(s"Deleted partition $tp from metadata cache in response to UpdateMetadata " +
-              s"request sent by controller $controllerId epoch $controllerEpoch with correlation id $correlationId")
+            if (traceEnabled)
```

Comment:
Should we add a comment that these trace logs are expensive?

Reply:
Yes, we can add a comment.